airbyte-integrations/bases/source-test-lib/src/main/java/io/airbyte/integrations/base/TestSource.java

```diff
@@ -0,0 +1,265 @@
+    assertTrue(output.getOutput().isPresent());
+  }
+
+  /**
+   * Verify that when given valid credentials, that check connection returns a success response.
+   * Assume that the {@link TestSource#getConfig()} is valid.
+   */
+  @Test
+  public void testCheckConnection() throws Exception {
+    final OutputAndStatus<StandardCheckConnectionOutput> output = runCheck();
+    assertTrue(output.getOutput().isPresent());
+    assertEquals(Status.SUCCEEDED, output.getOutput().get().getStatus());
+  }
+
+  // /**
+  // * Verify that when given invalid credentials, that check connection returns a failed response.
+  // * Assume that the {@link TestSource#getFailCheckConfig()} is invalid.
+  // */
+  // @Test
+  // public void testCheckConnectionInvalidCredentials() throws Exception {
+  // final OutputAndStatus<StandardCheckConnectionOutput> output = runCheck();
+  // assertTrue(output.getOutput().isPresent());
+  // assertEquals(Status.FAILED, output.getOutput().get().getStatus());
+  // }
+
+  /**
+   * Verify that when given valid credentials, that discover returns a valid catalog. Assume that the
+   * {@link TestSource#getConfig()} is valid.
+   */
+  @Test
+  public void testDiscover() throws Exception {
+    final OutputAndStatus<StandardDiscoverCatalogOutput> output = runDiscover();
+    assertTrue(output.getOutput().isPresent());
+    // the worker validates that it is a valid catalog, so we do not need to validate again (as long as
+    // we use the worker, which we will not want to do long term).
+    assertNotNull(output.getOutput().get().getCatalog());
+  }
+
+  /**
+   * Verify that the integration successfully writes records. Tests a wide variety of messages and
+   * schemas (aspirationally, anyway).
+   */
+  @Test
+  public void testRead() throws Exception {
+    final List<AirbyteMessage> recordMessages = runRead(getCatalog()).stream().filter(m -> m.getType() == Type.RECORD).collect(Collectors.toList());
+    // the worker validates the messages, so we just validate the message, so we do not need to validate
+    // again (as long as we use the worker, which we will not want to do long term).
+    assertFalse(recordMessages.isEmpty());
+  }
+
+  /**
+   * Verify that the integration overwrites the first sync with the second sync.
+   */
+  @Test
+  public void testSecondRead() throws Exception {
+    final List<AirbyteMessage> recordMessagesFirstRun =
+        runRead(getCatalog()).stream().filter(m -> m.getType() == Type.RECORD).collect(Collectors.toList());
+    final List<AirbyteMessage> recordMessagesSecondRun =
+        runRead(getCatalog()).stream().filter(m -> m.getType() == Type.RECORD).collect(Collectors.toList());
+    // the worker validates the messages, so we just validate the message, so we do not need to validate
+    // again (as long as we use the worker, which we will not want to do long term).
+    assertFalse(recordMessagesFirstRun.isEmpty());
+    assertFalse(recordMessagesSecondRun.isEmpty());
```

Comment:
this is susceptible to a race condition if the data changes in between the first and second run. we can rethink this if it's a problem in practice.